src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTestBase.java

```diff
@@ -50,30 +48,41 @@
   protected static final TopicPartition TOPIC_PARTITION2 = new TopicPartition(TOPIC, PARTITION2);
   protected static final TopicPartition TOPIC_PARTITION3 = new TopicPartition(TOPIC, PARTITION3);
 
+  protected static ElasticsearchContainer container;
   protected ElasticsearchClient client;
   private DataConverter converter;
 
+  @BeforeClass
+  public static void setupBeforeAll() {
+    // Relevant and available docker images for elastic can be found at https://www.docker.elastic.co
+    container = new ElasticsearchContainer("docker.elastic.co/elasticsearch/elasticsearch:7.0.0");
```

Comment:
These tests are currently unit tests, whereas this change would really warrant making them "integration" tests that require additional time to load the container. We normally run all of our integration tests during the Maven integration test phase, name them accordingly, and put them in the `integration` package. Should we make these test class integration tests instead?

My concern with doing this is that a lot of our unit tests are going to switch to be integration tests, and that our unit test coverage becomes pretty spartan.

An alternative approach is to *add* integration tests that effectively do what these tests do (and preferably more), yet keep these simple unit tests for faster compile-build-test cycles during development. Ideally we'd be able to run integration tests against several versions.

Reply:
Hi @rhauch this is actually a very valid question and I was expecting you to drop it ;-). I see this question in two sides, I will try to argument this here, hopefully in a clear way.

## Current state
As of now the connector test such as the ones inheriting from ElasticsearchSinkTestBase are having a direct interaction with an embedded Elasticsearch version. As you can see from the same class definition:

```java
public class ElasticsearchSinkTestBase extends ESIntegTestCase {
```

As of now this means, MappingTest, ElasticsearchSinkTaskTest and ElasticsearchWriterTest.

Taking as example _MappingTest_ for example we can see a mixture of test, for example 

```java
 public void testMapping() throws Exception {
    InternalTestCluster cluster = ESIntegTestCase.internalCluster();
    cluster.ensureAtLeastNumDataNodes(1);
```
is testing the creation and retrieval of a mapping with an embedded real Elasticsearch. This function is clearly an integration test as it will depend on the target system API and this is subject to changes across major versions.

we've as well
```java
  @Test
  @SuppressWarnings("unchecked")
  public void testStringMappingForES6() throws Exception {
    ElasticsearchClient client = mock(ElasticsearchClient.class);
    when(client.getVersion()).thenReturn(ElasticsearchClient.Version.ES_V6);

    Schema schema = SchemaBuilder.struct().name("textRecord")
            .field("string", Schema.STRING_SCHEMA)
            .build();
    ObjectNode mapping = (ObjectNode) Mapping.inferMapping(client, schema);
```

that is more a unit test per definition.

the problem with the first one is that makes part of our test dependant of ES internal versions. When code is develop like this, as is a common practise in the Elasticsearch world (aka no dependency management), plugins are released for specific versions. This is clearly not what we are aiming to do or need to do.

Other test classes mention here have a similar pattern. 

## How I envision things.

As we can't have a connector release for each Elasticsearch version, we would need to offer a more clear way to test between major versions. For this my proposal is to:

* Move as integration test, under integration test the ones that makes sense. For example test for creating and retrieving mappings as I commented earlier. For the tests move into integration test, because they are end of the day integration test, we should introduce mocks to abstract the end system call.
* Have a flexible integration test layer were we can pass a variable and pull different old and upcoming Elasticsearch versions. This will allow us to build a more resilient test matrix and detect support for upcoming or compatibility problems easier.

Does this makes sense? what do you think @rhauch and @kkonstantine ?

Reply:
Hi @kkonstantine, don't get me wrong, I'm not asking for having less unit test and more unit test. I'm actually advocating for better unit testing than what we have now.

End of the day what we've now is a mixture of integration test, "shaded" as unit test, this is because end of the day they go to a real elasticsearch, even if embedded. The way I've got to learn and apply unit test in my experience as developer, they never had interaction with an external system, this sort of defuse the idea of a unit test.

But I'm for sure open to change my mind, talking about the options presented in other messages, sorry if I'm missing any:

+ Mock out the interaction with the Elasticsearch server: This is my preferred option as it will enable proper entity test without interference of external apis. The integration with the external apis will be tested with the integration tests.
+ Work with an embed of a given version (probably latest): This will certainly safe a lot of need of mocks, but still all unit test code will be dependant on the last available version used of elasticsearch. I'm not confortable with this, but might be a solution to avoid having a ton of mocks.
+ Provide a parametrised version for an embed test:  I would certainly not enforce this, i don't even know how to start with it as the multiple versions and provided test. If possible I would discard this option.

What do you think @kkonstantine ?

Reply:
Hi @kkonstantine, I created https://github.com/confluentinc/kafka-connect-elasticsearch/pull/323 this weekend, hopefully it able to match if second point is more of your preference.

If this is the direction you would recommend this PR to take, I would kindly close this PR and move all the work there. 

I gave a bit of thought and I think moving forward implementing option 2 of the previous list is the most sensible way now, but certainly i would recommend improving the way unit tests are approached because the current approach is not providing enough level of confidence. Hopefully last is corrected with better real integration test, so end of the day we could remove the need of the real ES instance for the unit test. Makes sense?

Reply:
Talking about real examples of integration test that lay inside the current unit test suit, what do you think?

```java
  @Test
  public void testPutAndFlush() throws Exception {
    Map<String, String> props = createProps();

    ElasticsearchSinkTask task = new ElasticsearchSinkTask();
    task.start(props, client);
    task.open(new HashSet<>(Arrays.asList(TOPIC_PARTITION, TOPIC_PARTITION2, TOPIC_PARTITION3)));

    String key = "key";
    Schema schema = createSchema();
    Struct record = createRecord(schema);

    Collection<SinkRecord> records = new ArrayList<>();
    SinkRecord sinkRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, key, schema, record, 0);
    records.add(sinkRecord);

    sinkRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, key, schema, record, 1);
    records.add(sinkRecord);

    task.put(records);
    task.flush(null);

    client.refresh();

    verifySearchResults(records, true, false);
  }
```

the fact that we use _verifySearchResults_ it forces an interaction with the external system, at less based on my experience, a more suited unit test style would be, something on the sense of:

```java
```java
  @Test
  public void testPutAndFlush() throws Exception {
    Map<String, String> props = createProps();

    ElasticsearchSinkTask task = new ElasticsearchSinkTask();
    task.start(props, client);
    task.open(new HashSet<>(Arrays.asList(TOPIC_PARTITION, TOPIC_PARTITION2, TOPIC_PARTITION3)));

    String key = "key";
    Schema schema = createSchema();
    Struct record = createRecord(schema);

    Collection<SinkRecord> records = new ArrayList<>();
    SinkRecord sinkRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, key, schema, record, 0);
    records.add(sinkRecord);

    sinkRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, key, schema, record, 1);
    records.add(sinkRecord);

    task.put(records);
    task.flush(null);

   // in mockito or other similar framework call
  // verify executeBulk is called the exact number of times, that for this test should be 1
 
  }
```

this would be the approach I propose to take for the first point in the previous list. While the current approach "solve the problem", it makes the test a bit harder to trust as they depend on a given elasticsearch version. 

What do you think?

Reply:
IMO, we should not get too hung up on whether "unit" tests never hit an external system and that "integration" tests always hit an external system. IIUC, the embedded version of Elasticsearch was intended primarily for testing purposes so it becomes easier to test individual components and logic, and in that sense I don't think of the embedded Elasticsearch as a "real" external system, and is closer to a "mock" system. In fact, I think we should avoid creating our own "mock" system because isn't worth the effort or risk. Let's not be pedantic about what is a "unit" vs "integration" tests. I'd much rather be pragmatic.

I do think it's acceptable to have "unit" tests that use a single embedded version of Elasticsearch. Granted, we should probably use a more recent version than what we've been using, but the purpose of these unit tests is to verify basic function and operation _of individual components and logic_, not of the entire connector. The point is that they are quick and focused on testing *individual components*. If it happens to be easiest to verify this using a real embedded Elasticsearch instance, then so be it.

Those "unit" tests would not really cover compatibility with any versions other than the one selected version, and this is where the ITs come in with two main purposes:
a) run the whole connector via Connect distributed worker(s) to validate the lifecycle and verify proper functionality
b) do this against a series of Elasticsearch versions (e.g., the latest patch on multiple minor versions)

Docker (or Test Containers) may be the most effective way to satisfy b) while running *real* instances of Elasticsearch. One way to do this with Docker and the Fabric8 plugin is to have different profiles for different versions of Elasticsearch, such that I could use `-Pv7.2.0` (for example) run the same ITs against the 7.2.0 version (other versions could have similar profiles). ([Elastic's Docker registry](https://www.docker.elastic.co/) contains nearly every version, unlike Docker Hub's.) Locally I could then run `-Pv7.2.0,v6.8.1` to sequentially run the ITs against multiple versions in one build. Once we get ITs runnable on Jenkins, we could change the `-Pjenkins` profile to run multiple of these profiles this way.

Test Containers gives us a bit more flexibility without all the profiles, since we could create a separate test suite (classes) for each version we wish to test. A base class would contain all of the real tests, but the concrete subclasses would specify the version of ES that would be used for those tests.

Reply:
BTW, I think this also is aligned with @kkonstantine's previous second bullet, which @purbon also seemed to favor. I think this also allows us to do this in steps: first get ES 7 working with the connector, then (in a separate PR) add ways of running ITs against multiple ES versions for compatibility testing. We could improve the unit testing (as you suggested above, @purbon) to rely less upon an embedded ES server in yet another subsequent PR, though as I stated above, IMO that is lower priority.